Bond token (#413)
scale-it/algo-builder: examples/bond/README.md

- We bundle coupons with bonds in a single token - Bond token (ASA). 
    - There is a different ASA for each epoch: in a single `BondApp` we define family of Bond tokens: `B_0`, `B_1`, `B_2` , … `B_n` .  `B_i` token represent a bond which received a coupon payment at `i-1` epoch but execute a coupon payment at `i` epoch.
    - Holder of `B_i` token can redeem his token and get coupon payment using DEX lsigs (see below)
- Set of logical signatures (stateless contracts): `DEX_0`, `DEX_1` ….
    - `B_i` holder can exchange tokens with `DEX_i` for equal number of `B_{i+1}` tokens and `coupon_value` of ALGOs only when `i < BondApp.epoch`. For example: At epoch 2, user X owns 10 `B_1`  tokens. He can send his tokens to `DEX_1` and in exchange of 10 `B_2` tokens and `10 × BondApp.coupon_value` of ALGOs. NOTE: at epoch 2, user can exchange his `B_0` tokens with `DEX_0` if he didn’t do it yet.
    -  `B_i` tokens sent to `DEX_i` are locked forever. 

>     - Issuer must create and supply ALGO to `DEX_i` in advance. Otherwise legal consequences can be triggered.

Author: ok

# Algorand bond token

Bonds are units of corporate debt issued by companies and securitized as tradeable assets. A bond is referred to as a fixed-income instrument since bonds traditionally paid a fixed interest rate (coupon) to debt holders. Variable or floating interest rates are also now quite common. Bond prices are inversely correlated with interest rates: when rates go up, bond prices fall and vice-versa. 

In this template, we use ALGO to express the nominal and trading values (investors will buy, sell and exchange bonds using ALGO). It's easy to change this and use any other ASA.

When creating a bond the following information must be recorded:

> - `issue_price` — price at which bonds are sold by the issuer.

Reviewer: ```suggestion
- `issue_price` — price of a bonds sold by the issuer.
```

- `exit(amount)`  — after maturity bond holder can send bonds back to the issuer and receive `amount×nominal_price`  back. Issuer must provide enough funds to the issuer account upon the maturity date. Otherwise a legal consequences can be triggered.
- `redeem_coupon()`  — bond holder can get his interest (in ALGO).

Note: In this example we have constant `coupon_value` for all the epochs, we can make variable by having this value hardcoded in each `DEX_i`.

## Spec document

App template [specification](https://paper.dropbox.com/doc/Algorand-Bond-Template--BOU8bTQSnmRNk23KK8McWwxXAg-hzI7C681Soo2sr6iyGFzg).

## Deploy script

We have deploy script in `scripts/deploy`, This script deploys initial bond token, deploys bond dapp application and updates issuer address in bond-dapp contract.

## Run script

In `scripts/run` folder we have scripts with the following files:

 - `run/run.js`: It controls all the scripts and maintain order of execution.
 - `run/issue.js`: This is first script, in this script tokens are issued to issuer from token creator.
 - `run/epoch0.js`: In this script, `elon` buys 10 bond tokens and sell 2 bond tokens to `bob` for 2020 Algos.

>  - `run/createDex.js`: This script creates ith Dex, burn B_i bond tokens, issue B_i+1 bond tokens.

Reviewer: ```suggestion
 - `run/createDex.js`: This script creates i-th Dex, burns `B_i` bond tokens, issues `B_i+1` bond tokens.
```

>  - `run/redeem.js`: This script redeems tokens from a given dex id.

Reviewer: ```suggestion
 - `run/redeem.js`: This script exchanges tokens and redeems coupon.
```

 - `run/epoch1.js`: `elon` redeems 8 bond tokens and buys 4 more from dex.
 - `run/createBuyback.js`: Creates a buyback logic signature address.
 - `run/exit.js`: Buyer can exit their bond tokens using this script.
 - `run/common/accounts.js`: Loads accounts
 - `run/common/common.js`: contains common functions and constants.
 

>  - In run script we have the following scenario:

Reviewer: ```suggestion
 - In a run script example we present the following scenario:
```

>     - Issue initial bond tokens to issuer

Reviewer: ```suggestion
    - Issue initial bond tokens to the issuer
```

    - In epoch_0 elon buys 10 bonds
    - In epoch 0 elon sells 2 bonds to bob for 2020 ALGO (in a group transaction)
    - Manager creates dex 1

>     - Elon redeems her bonds(8), Elon buys 4 more bonds (so he will have 12 bonds in total)

Reviewer: ```suggestion
    - Elon redeems his bonds (8), Elon buys 4 more bonds (so he will have 12 bonds in total)
```

    - Manager creates dex 2

>     - Elon redeems all this bonds.

Reviewer: ```suggestion
    - Elon redeems all his bonds.
```

>     - Bob redeems his bonds_1 and then bonds_2 (so going through the previous dex)

Reviewer: ```suggestion
    - Bob redeems his bonds from epoch 0 and 1
```

>     - Maturity period is set to 240 seconds(4 min) from the time of deployment of contract. At maturity manager creates and funds buyback and both elon and bob exit all their tokens (12 and 2 respectively).

Reviewer: ```suggestion
    - Maturity period is set to 240 seconds(4 min) after the contract deployment. At maturity, manager creates and funds buyback and both elon and bob can exit all their tokens (12 and 2 respectively).
```
